File: sdk/python/controldb/integrations/opentelemetry.py
"""OpenTelemetry trace propagation helpers.

If ``opentelemetry-api`` is installed, ``current_trace_ids`` reads the active
span and returns the W3C trace context the SDK should attach to events. When
OTel is not installed, returns ``(None, None, None)``.
"""

from __future__ import annotations

from typing import Optional, Tuple


def current_trace_ids() -> Tuple[Optional[str], Optional[str], Optional[str]]:
    try:
        from opentelemetry import trace  # type: ignore
    except Exception:  # pragma: no cover - OTel optional
        return None, None, None
    span = trace.get_current_span()
    context = span.get_span_context() if span else None
    if context is None or not context.is_valid:
        return None, None, None
    trace_id = format(context.trace_id, "032x")
    span_id = format(context.span_id, "016x")
    return trace_id, span_id, None

File: sdk/typescript/src/integrations/opentelemetry.ts
/**
 * OpenTelemetry trace propagation helpers.
 *
 * If @opentelemetry/api is installed, currentTraceIds() reads the active span
 * and returns the W3C trace context IDs the SDK should attach to events.
 * When OTel is not installed, returns an empty object.
 *
 * Mirrors sdk/python/controldb/integrations/opentelemetry.py exactly.
 */

export interface TraceIds {
  traceId?: string;
  spanId?: string;
  parentSpanId?: string;
}

/**
 * Returns trace/span IDs from the active OpenTelemetry span, if available.
 * Falls back gracefully to an empty object when @opentelemetry/api is not installed.
 */
export async function currentTraceIds(): Promise<TraceIds> {
  try {
    // Dynamic import so @opentelemetry/api remains optional
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const otelApi = await import("@opentelemetry/api");
    const span = otelApi.trace.getActiveSpan();
    if (!span) return {};
    const context = span.spanContext();
    if (!context || !otelApi.isSpanContextValid(context)) return {};

    const traceId = context.traceId; // already 32-char hex string
    const spanId = context.spanId;   // already 16-char hex string
    return { traceId, spanId };
  } catch {
    // OTel not installed or no active span
    return {};
  }
}

/**
 * Synchronous variant — returns an empty object if OTel is not available.
 * Use this when you cannot await.
 */
export function currentTraceIdsSync(): TraceIds {
  try {
    // Synchronous require (CommonJS only; works in Node.js)
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const otelApi = require("@opentelemetry/api") as typeof import("@opentelemetry/api");
    const span = otelApi.trace.getActiveSpan();
    if (!span) return {};
    const context = span.spanContext();
    if (!context || !otelApi.isSpanContextValid(context)) return {};
    return { traceId: context.traceId, spanId: context.spanId };
  } catch {
    return {};
  }
}
